### PR TITLE
Initialize Bolt PPC Button

### PR DESCRIPTION
**Pull request by var-bolt**
This PR implements a basic version of PPC functionality. The goal is to be able to test the end to end flow of PPC.

I have uploaded the cartridge to https://zzgv-010.dx.commercecloud.salesforce.com/s/RefArch/
Please go to any product page and change the product attributes so show / hide the PPC button. You should be able to checkout successfully via PPC button.

Pending issues:
1) Show / hide PPC button function is invoked in response to a `product:updateAddToCart` event. So unless the user updates the product attributes it won't be invoked -> which means PPC button won't be shown unless user changes product attributes
2) Similar to point 1. We need to find a way to show PPC button on page load if product is ready to be ordered. Certain products are in the ready to order state as soon as the page loads. The existing implementation won't show PPC unless user changes product attributes
3) Current implementation won't work on pages that have product bundles or product sets
4) Current implementation won't include priced attributes. Eg: Warranty on page
   Eg: See this page: https://zzgv-010.dx.commercecloud.salesforce.com/s/RefArch/electronics/digital%20cameras/kodak-z8612M.html?lang=en_US
   Warranty price won't be included in the product page checkout cart.
5) Order confirmation does not appear to load
6) [Involves arch design] How do we associate order to user if they are logged into the store?

Note:
- Will remove certain comments and console statements after a couple rounds of review. Want to get initial feedback first

### CODE REVIEW

File: cartridges/int_bolt_sfra/cartridge/client/default/js/boltProductPageButton.js [hunk 1]
```diff
@@ -1,22 +1,42 @@
...
         if (boltButtonLoaded) {
             clearInterval(boltButtonExist);
             console.log('Bolt connect has loaded and product page button is ready!');
         }
     }, 250);
 };
 
+var callBoltConfigure = function (product) {
+    if (product.available && product.readyToOrder) {
+        console.log('Ready to call BoltCheckout.Configure');
```
Comment (dumega) [inline cartridges/int_bolt_sfra/cartridge/client/default/js/boltProductPageButton.js:21]: @var-bolt Is this for debugging only? I do not see any logic in this function.

Comment (var-bolt) [reply cartridges/int_bolt_sfra/cartridge/client/default/js/boltProductPageButton.js]: I have since updated the PR! Please take a second look 😄

**Comment by dumega** [comment]
For issues 1&2

There are two solutions:
1) We can first create an endpoints with a controller which returns the current product object. When the product details page is loaded, we can fetch the product data via ajax call for initiate Bolt PPC button.
2) This solution would be a bit hacky, I have tried it on my end and it does work. So once the page is loaded, we force to trigger the event change on quantity select element which would start the process to initiate Bolt PPC button. Below is the code snippet:

```
$(document).ready(function () {
    var quantityButtonExist = setInterval(function () {
        var quantityButton = $('.quantity-select');
        var buttonLoaded = quantityButton && quantityButton.is(":visible");
        if (buttonLoaded) {
            clearInterval(quantityButtonExist);
            quantityButton.change()
        }
    }, 100);
    // product:statusUpdate
    $(document).bind('product:updateAddToCart', function (_e, productResponse) {
        console.log('On Status Update', productResponse);
        var product = productResponse.product;
        if (product.available && product.readyToOrder) {
            var boltCartObject = buildBoltCartObject(product);
            console.log(boltCartObject);
            configurePPCCartAndShowButton(boltCartObject);
        } else {
            clearPPCCartAndHideButton();
        }
    });
});
```

For issues 3&4

Now the cartridge sends product id and quantity to the Bolt server, so Bolt does not have any details of product bundles or priced attributes. If you want to support product bundles and priced attributes, I think you need to add those data to BoltCartObject as well


For issue 5, the order confirmation page does not redirect correctly, it is because the Bolt checkout transaction in callback which is returned from Bolt server  does not contain merchant_order_number, please check my screenshot. I think you need to look into your code to fix this issue.
image


For issue 6 How do we associate order to user if they are logged into the store?
I think we discuss this before, for logged-in customer, we can get his customer_number and then add this number to customer_info field of api for creating basket, for guest, just provide email.

**Comment by dumega** [comment]
@var-bolt  Please check https://github.com/BoltApp/bolt-demandware-managed/pull/68 and it contains the solutions for issues #1 #2 and #4

**Pull request merged**